internal/provider/datasources/deployment.go

```diff
@@ -0,0 +1,405 @@
+	"github.com/prefecthq/terraform-provider-prefect/internal/api"
+	"github.com/prefecthq/terraform-provider-prefect/internal/provider/customtypes"
+	"github.com/prefecthq/terraform-provider-prefect/internal/provider/helpers"
+	"github.com/prefecthq/terraform-provider-prefect/internal/provider/resources"
+)
+
+// Ensure the implementation satisfies the expected interfaces.
+var (
+	_ datasource.DataSource              = &deploymentDataSource{}
+	_ datasource.DataSourceWithConfigure = &deploymentDataSource{}
+)
+
+// deploymentDataSource is the data source implementation.
+type deploymentDataSource struct {
+	client api.PrefectClient
+}
+
+// DeploymentDataSourceModel defines the Terraform data source model.
+type DeploymentDataSourceModel struct {
+	// The model requires the same fields, so reuse the fields defined in the resource model.
+	resources.DeploymentResourceModel
+}
+
+// NewDeploymentDataSource is a helper function to simplify the provider implementation.
+//
+//nolint:ireturn // required by Terraform API
+func NewDeploymentDataSource() datasource.DataSource {
+	return &deploymentDataSource{}
+}
+
+// Metadata returns the data source type name.
+func (d *deploymentDataSource) Metadata(_ context.Context, req datasource.MetadataRequest, resp *datasource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_deployment"
+}
+
+// Schema defines the scema for the data source.
+func (d *deploymentDataSource) Schema(_ context.Context, _ datasource.SchemaRequest, resp *datasource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		Description: `
+Get information about an existing Deployment by either:
+- deployment ID, or
+- deployment name
+The Deployment ID takes precedence over deployment name.
+`,
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
+				Computed:    true,
+				Description: "Deployment ID (UUID)",
+				Optional:    true,
+			},
+			"created": schema.StringAttribute{
+				Computed:    true,
+				CustomType:  customtypes.TimestampType{},
+				Description: "Timestamp of when the resource was created (RFC3339)",
+			},
+			"updated": schema.StringAttribute{
+				Computed:    true,
+				CustomType:  customtypes.TimestampType{},
+				Description: "Timestamp of when the resource was updated (RFC3339)",
+			},
+			"account_id": schema.StringAttribute{
+				CustomType:  customtypes.UUIDType{},
+				Description: "Account ID (UUID), defaults to the account set in the provider",
+				Optional:    true,
+			},
+			"workspace_id": schema.StringAttribute{
+				CustomType:  customtypes.UUIDType{},
+				Description: "Workspace ID (UUID) to associate deployment to",
+				Optional:    true,
+			},
+			"name": schema.StringAttribute{
+				Computed:    true,
+				Description: "Name of the deployment",
+				Optional:    true,
+			},
+			"flow_id": schema.StringAttribute{
+				CustomType:  customtypes.UUIDType{},
+				Description: "Flow ID (UUID) to associate deployment to",
+				Optional:    true,
```

Comment:
aside from the user-provided inputs (name, id, account_id, workspace_id), should any attribute be `Optional` and not just read-only `Computed: true`?

Reply:
good call - I understand this better now, updated in 1b35a3d (but let me know if that doesn't match what you were thinking)